p2panda-rs/src/schema/system_schema.rs

```diff
@@ -0,0 +1,327 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+use std::convert::TryFrom;
+use std::str::FromStr;
+
+use crate::document::{DocumentView, DocumentViewId};
+use crate::operation::{OperationValue, Relation};
+
+use super::SystemSchemaError;
+
+#[derive(Clone, Debug, Copy, PartialEq)]
+#[allow(missing_docs)]
+pub enum FieldType {
+    Bool,
+    Int,
+    Float,
+    String,
+    Relation,
+}
+
+impl FromStr for FieldType {
+    type Err = SystemSchemaError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "bool" => Ok(FieldType::Bool),
+            "int" => Ok(FieldType::Int),
+            "float" => Ok(FieldType::Float),
+            "str" => Ok(FieldType::String),
+            "relation" => Ok(FieldType::Relation),
+            _ => Err(SystemSchemaError::InvalidFieldType),
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct SchemaView {
+    // ID of this schema view.
+    id: DocumentViewId,
+    /// Name of this schema.
+    name: String,
+    /// Description of this schema.
+    description: String,
+    /// The fields in this schema.
+    /// TODO: This will be a `relation` when we merge that PR
+    fields: RelationList,
+}
+
+type RelationList = Vec<Relation>;
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct SchemaFieldView {
+    // ID of this schema field view.
+    id: DocumentViewId,
+    /// Name of this schema field.
+    name: String,
+    /// Type of this schema field.
+    field_type: FieldType,
+}
+
+/// View onto materialised schema which has fields "name", "description" and "fields".
+/// Is validated on being converted from a general DocumentView struct which means so it's inner
+/// values can be returned unwrapped by their getter methods.
```

Comment:
This sentence is somehow hard to read 😅 maybe it should say something like: "the fields are validated when converting a DocumentView struct into this type".

Also a newline after the first sentence will make the docstring look nicer in HTML 👍

Reply:
Ah, this is a zombie comment.... will fix it up nice.